core/consensus/babe/impl/babe_impl.cpp

```diff
@@ -522,51 +661,54 @@ namespace kagome::consensus {
+        if (not first_production_slot) {
+          first_production_slot = observed_slot + kSlotTail;
+          log_->info("Peer will start produce blocks at slot number: {}",
+                     *first_production_slot);
+        }
 
-    // get the difference between observed slot and the one that we are trying
-    // to launch
-    auto diff = *first_production_slot - observed_slot;
+        if (observed_slot < first_production_slot.value()) {
+          storeFirstSlotTimeEstimate(observed_slot, *first_production_slot);
+          return;
+        }
 
-    first_slot_times_.emplace_back(
-        clock_->now() + diff * genesis_configuration_->slot_duration);
-    if (observed_slot >= first_production_slot.value()) {
-      current_state_ = State::SYNCHRONIZED;
-      log_->info("Slot time obtained. Peer is synchronized");
+        current_state_ = State::SYNCHRONIZED;
+        log_->info("Slot time obtained. Peer is synchronized");
 
-      // get median as here:
-      // https://en.cppreference.com/w/cpp/algorithm/nth_element
-      std::nth_element(first_slot_times_.begin(),
-                       first_slot_times_.begin() + first_slot_times_.size() / 2,
-                       first_slot_times_.end());
-      BabeTimePoint first_slot_ending_time =
-          first_slot_times_[first_slot_times_.size() / 2];
+        auto first_slot_ending_time = getFirstSlotTimeEstimate();
 
-      Epoch epoch;
-      epoch.epoch_index =
-          *first_production_slot / genesis_configuration_->epoch_length;
-      epoch.start_slot = *first_production_slot;
-      epoch.epoch_duration = genesis_configuration_->epoch_length;
-      auto next_epoch_digest_res =
-          epoch_storage_->getEpochDescriptor(epoch.epoch_index);
-      if (not next_epoch_digest_res) {
-        log_->error("Could not fetch epoch descriptor for epoch {}. Reason: {}",
-                    epoch.epoch_index,
-                    next_epoch_digest_res.error().message());
-        return;
-      }
-      epoch.randomness = next_epoch_digest_res.value().randomness;
-      epoch.authorities = next_epoch_digest_res.value().authorities;
-
-      runEpoch(epoch, first_slot_ending_time);
+        epoch = prepareFirstEpochFromZeroStrategy(first_slot_ending_time,
+                                                  *first_production_slot);
 
-      if (auto on_synchronized = std::move(on_synchronized_)) {
-        on_synchronized();
+        runEpoch(epoch, first_slot_ending_time);
+        break;
       }
+      case SlotsStrategy::FromUnixEpoch: {
+        auto last_known_epoch = epoch_storage_->getLastEpoch().value();
+        epoch = prepareFirstEpochUnixTime(last_known_epoch,
+                                          babe_header.slot_number + 1);
+
+        {
+          // calculate new epoch first slot finish time and run epoch
+          auto last_epoch_starting_slot = last_known_epoch.start_slot;
+          auto last_epoch_first_epoch_starting_time =
```

Comment:
Is name of variable correct?

Reply:
should be `last_epoch_first_slot_starting_time `. Thanks for notice